File: 106_construct_binary_tree_from_inorder_and_postorder_traversal.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.empty()) return NULL;
        return dfs(inorder,postorder,0,inorder.size()-1,0,postorder.size()-1);
    }
    TreeNode* dfs(vector<int>& inorder, vector<int>& postorder, int s1, int e1,int s2,int e2) {
        if (s1>e1) return NULL;
        TreeNode* root=new TreeNode(postorder[e2]);
        int index=s1;
        for (;index<=e1;index++)
            if (inorder[index]==postorder[e2]) break;
        root->left=dfs(inorder,postorder,s1,index-1,s2,s2+index-1-s1);
        root->right=dfs(inorder,postorder,index+1,e1,s2+index-s1,e2-1);
        return root;
    }
};

File: algorithms/python3/106_construct_binary_tree_from_inorder_and_postorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def helper(inorder, postorder):
            if not inorder:
                return None
            root = postorder.pop()
            head = TreeNode(root)
            ind = inorder.index(root)
            head.left = helper(inorder[:ind], postorder[:ind])
            head.right = helper(inorder[ind + 1:], postorder[ind:])
            return head

        return helper(inorder, postorder)
